artiq/test/coredevice/test_moninj.py

```diff
@@ -66,3 +71,237 @@ def injection_status_cb(channel, override, value):
+        channel.set(freq, phase)
+
+    @kernel
+    def write_with_amp(self, channel, freq: TFloat, phase: TFloat = 0.0, amplitude: TFloat = 0.0):
+        self.init_channel(channel)
+        channel.set(freq, phase, amplitude)
+
+
+class AD991XMonitorTest(ExperimentCase, IsolatedAsyncioTestCase):
+    def get_urukuls(self):
+        urukuls = defaultdict(dict)
+        ddb = self.device_db.get_device_db()
+        for name, desc in ddb.items():
+            if isinstance(desc, dict) and desc["type"] == "local":
+                module, cls = desc["module"], desc["class"]
+                if (module, cls) == ("artiq.coredevice.ad9910", "AD9910"):
+                    urukuls[cls][name] = self.device_mgr.get(name)
+                if (module, cls) == ("artiq.coredevice.ad9912", "AD9912"):
+                    urukuls[cls][name] = self.device_mgr.get(name)
+        return urukuls
+
+    def urukuls_all(self):
+        ret = {}
+        for x in self.urukuls.values():
+            ret.update(x)
+        return ret
+
+    def ensure_ad9910_only(self):
+        if len(self.urukuls["AD9910"]) < 1:
+            raise unittest.SkipTest("test device not available: no ad9910 devices")
+
+    def ensure_ad9912_only(self):
+        if len(self.urukuls["AD9912"]) < 1:
+            raise unittest.SkipTest("test device not available: no ad9912 devices")
+
+    def setUp(self):
+        super().setUp()
+        self.core = self.device_mgr.get_desc("core")
+        self.urukuls = self.get_urukuls()
+        if len(self.urukuls) < 1:
+            raise unittest.SkipTest("test device not available: no urukul devices")
+        self.kernel = self.create(_UrukulExperiment)
+
+    @contextmanager
+    def open_comm_session(self, loop, notifications=None, core_host=None, auto_monitor=True):
+        if notifications is None:
+            notifications = []
+        if core_host is None:
+            core_host = self.core["arguments"]["host"]
+
+        def monitor_cb(channel, probe, value):
+            notifications.append((channel, probe, value))
+
+        moninj_comm = CommMonInj(monitor_cb, lambda x, y, z: None)
+        try:
+            loop.run_until_complete(moninj_comm.connect(core_host))
+            if auto_monitor:
+                self.setup_monitor(moninj_comm)
+            yield moninj_comm
+        finally:
+            loop.run_until_complete(moninj_comm._writer.drain())
+            loop.run_until_complete(moninj_comm.close())
+
+    def setup_monitor(self, moninj_comm, *, reset_zero=True):
+        for name, dev in self.urukuls_all().items():
+            moninj_comm.monitor_probe(True, dev.bus.channel, dev.chip_select - 4)
+            if reset_zero:
+                self.kernel.write_raw(dev, 0)
+
+    def test_double_read(self):
+        target_ftw = 0xff00ff00
+        notifications_out = []
+
+        loop = asyncio.new_event_loop()
+        try:
+            with self.open_comm_session(loop, notifications_out):
```

Comment:
Doesn't unittest provide features to deal with that boilerplate? Look at hardware_testbench.py. You may also want to extend it and convert the other moninj tests to that (in a separate PR).

Reply:
done, let's be easy and put it in setUp/tearDown which is what every unit test runs before it start/stop